Assets/Stack mini game/Scripts/GameplayManager.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class GameplayManager : MonoBehaviour
+{
+    [SerializeField] private CubeSpawner cubeSpawner;
+
+    [SerializeField] private CubeStack cubeStack;
+
+    [SerializeField] private int fullWinScore;
+
+    [SerializeField] private int partialWinScore;
+
+    [SerializeField] private GameState gameStates;
+
+    public GameState GameStates => gameStates;
+
+    public Action GameResetEvent;
+
+    public Action GameEndEvent;
+
+
+    private void Start()
+    {
+        ResetGame();
+    }
+
+    // When the player presses the play button, he or she attempts to stack the current moving cube on top of the cube below
+    // If the cubes are stacked, the game goes on, and we spawn a new cube on top
+    // If not it's game over
+
+    public void DoPlayButton()
+    {
+        CubeCornersPositionTracker topCubeCornerPosition = cubeStack.CubeCornersPositionPile.CubeCornersPositionList[0];
+
+        // There is only one cube at the beginning. The player can place it wherever. There is no cutting involved yet.
+        if (cubeStack.CubeCornersPositionPile.CubeCornersPositionList.Count == 1)
+        {
+            cubeSpawner.SpawnStackedCube(spawnPosition: cubeStack.CubeCornersPositionPile.CubeCornersPositionList[0].transform.position,
+                                         setWidth: cubeStack.CubeCornersPositionPile.CubeCornersPositionList[0].transform.localScale.x);
+
+            return;
+        }
+
+        float cubeOverlapDistance = cubeStack.CubesOverlapDistance();
+        CubeCornersPositionTracker bottomCubeCornerPosition = cubeStack.CubeCornersPositionPile.CubeCornersPositionList[1];
+
+        // Cubes aren't stacked. It's game over
+        if (cubeOverlapDistance == 0f)
+        {
+            if(cubeStack.StackedCubes.Count >= partialWinScore+1)
+                gameStates.SetCurrentState("partial win");
+            else
+                gameStates.SetCurrentState("lose");
+
+            GameEndEvent?.Invoke();
```

Comment:
We might want an event that passes the State as a parameter. We could have the state manager as a singleton in the main project though instead but if we’re using events anyway we may as well send the state through there. Either way is fine. Or we could have a FullWin event, PartWin etc. and GameLost.

Either way we will in the main project need to use the game state elsewhere, to repair the ship and expend repair tools etc.